### PR TITLE
[VFX] New ShaderGraph integration support in URP

### PR DESCRIPTION
**Pull request by PaulDemeulenaere**
---
### Purpose of this PR
This PR follows up the initial new integration for HDRP : https://github.com/Unity-Technologies/Graphics/pull/3467
Add support of the new ShaderGraph integration with VFX in URP.

- `VFXHDRPSubTarget.cs` has been deleted and factorized into `VFXSubTarget.cs`, the specific SRP description is provided by the `VFXSRPBinder`.
- Moved `ConfigPlanarPrimitive.template.hlsl`, `ConfigMesh.template.hlsl` & `Config.template.hlsl` to the VFX Graph package

### Target Support
It only supports the target:
- UniversalLitSubTarget
- UniversalUnlitSubTarget

And we don't support:
- UniversalDecalSubTarget
- UniversalSpriteLitSubTarget
- UniversalSpriteUnlitSubTarget
- UniversalSpriteCustomLitSubTarget

Additionally, VFX doesn't support the motion vector rendering in URP (the integration is only active in HDRP for now).

New Jira Tasks :
- [VFX Motion Vector in URP](https://jira.unity3d.com/browse/GVFXG-174)
- [SG integration on URP for SpriteLit/Unlit](https://jira.unity3d.com/browse/GVFXG-170)

See this [conversation](https://unity.slack.com/archives/C01Q3N39UEB/p1629821699001900).
_warning_unsupported

### CastShadow & MaterialOffset
The castShadow is controlled by the ShaderGraph in URP, add a user feedback help box:
_urp_warning
Additionally, the material offset settings was available in HDRP, the field `materialOffset` has been filtered out.
_hdrp_warning

### Special changes in HDRP
- [templateMaterialDirectories](https://github.com/Unity-Technologies/Graphics/pull/5448/files#diff-c9235421e1a8e3c7e1333c32b46dbb7aebe206f0bc260f3b16ca3919916d3ca5R24) has been factorized to easily add the vfx template path everywhere.
These changes have already been checked by @alelievr.

### Special changes in URP
- [BaseShadeGUI](https://github.com/Unity-Technologies/Graphics/pull/5448/files#diff-c2794c0b501c1cd6be0d600f6ee894b0eb7678f0aee6aebd5a8df555f8d7d6cbR244) integrates a material filter to easily filter out SurfaceInputs (equivalent behavior in HDRP)
- Factorization of the couple [BuildSurfafaceDescriptionInputs/SurfaceDescriptionFunction](https://github.com/Unity-Technologies/Graphics/pull/5448/files#diff-6563251b0ced4377543c5f59ba4492a5b63a02156eeeeda83dc23168fb1999e3L19) into BuildSurfaceDescription to ease VFX modification.
- Add [PostProcessSubShader](https://github.com/Unity-Technologies/Graphics/pull/5448/files#diff-2533d8d9accd6507e86de0442763e97c984880248eb96566b83b6d62ebbfd4f9R39) virtual function in UniversalSubTarget to alter subtarget if vfx graph support is enabled (equivalent behavior in HDRP)

### Special changes in VFX
- Rename Attributes structure in VFXAttributes (and SourceAttributes in VFXSourceAttributes) to resolve an annoying naming conflict (see this [conversation](https://unity.slack.com/archives/G1BTWN88Z/p1626962302013700))
- Fix [target selection](https://github.com/Unity-Technologies/Graphics/pull/5448/files#diff-7ca343b7be44f1b91bc00c69fc8bf8c9826dbd2bbba31cec77a34e0b284cb19bR748) while generating shader code
- Moved VFXSRPBinder to an isolated dedicated file

---
## Testing status
Added a graphicTest (which is a copy from the HDRP):
_parity_urp_hdrp

Tested manually the multiply blend mode:
image
See this [conversation](https://unity.slack.com/archives/C01Q3N39UEB/p1629985747005600).

Early test with vertex displacement:
_tomato_generator (2)

---
### Comments to reviewers
- See this [thread](https://unity.slack.com/archives/C01Q3N39UEB/p1629821527000600)
- See the [test document](https://docs.google.com/document/u/1/d/1OsM3iA2O4-Zb46eUUOZRFjMaVLmWFfMkG7xLAkpkah4/edit?usp=sharing)
- See the [JIRA](https://jira.unity3d.com/browse/GVFXG-123) card

Some PR which conflicts or conflicted with these changes:
- VFX_URP adds coverage on vulkan: https://github.com/Unity-Technologies/Graphics/pull/5436 ✔️ 
- Minor change in Quad Output common code: https://github.com/Unity-Technologies/Graphics/pull/5149 ✔️ 
- Fix integration of buffer/texture sampling: https://github.com/Unity-Technologies/Graphics/pull/5416 ⏳ 
- VFXURPBinder modifications: [#255 (internal)](https://github.cds.internal.unity3d.com/unity/vfx-graphics/pull/255) ⏳ 

TODO:
- [x] Update documentation (report same change than [HDRP](https://github.com/Unity-Technologies/Graphics/blob/5e335c0df9d04dabcbe29eb22352f7c5153496f0/com.unity.render-pipelines.high-definition/Documentation~/snippets/shader-properties/support-vfx-graph.md)) done at 5ccf5540ddd57c92a7f2db175ec8f78f52ca248b (see this [preview](https://github.com/Unity-Technologies/Graphics/blob/5ccf5540ddd57c92a7f2db175ec8f78f52ca248b/com.unity.shadergraph/Documentation~/surface-options.md))
- [x] Revert this [change](https://github.com/Unity-Technologies/Graphics/pull/5448/commits/61ebca16d50e3eec30164e9520d4a68ea02a584b), it will be include in https://github.com/Unity-Technologies/Graphics/pull/5475

### CODE REVIEW

File: com.unity.visualeffectgraph/Editor/ShaderGraph/Templates/VFXConfigPlanarPrimitive.template.hlsl [hunk 1]
```diff
@@ -8,7 +8,7 @@
 
 #define HAVE_VFX_PLANAR_PRIMITIVE
 
-bool GetMeshAndElementIndex(inout AttributesMesh input, inout AttributesElement element)
+bool GetMeshAndElementIndex(inout VFX_SRP_ATTRIBUTES input, inout AttributesElement element)
 {
     uint id = input.vertexID;
```
Comment (PaulDemeulenaere) [inline com.unity.visualeffectgraph/Editor/ShaderGraph/Templates/VFXConfigPlanarPrimitive.template.hlsl:13]: This file has been modified by this PR: https://github.com/Unity-Technologies/Graphics/pull/5149
The changes should have been correctly reported.
See these [specific lines](https://github.com/Unity-Technologies/Graphics/pull/5149/files#diff-d60723623314f24c8b46851360b97e4a3453b1a724d740c51b37a629bb3916bdR76)

File: com.unity.render-pipelines.universal/Editor/ShaderGUI/BaseShaderGUI.cs [hunk 2]
```diff
@@ -241,15 +241,30 @@ public override void OnGUI(MaterialEditor materialEditorIn, MaterialProperty[] p
             ShaderPropertiesGUI(material);
         }
 
+        protected virtual uint materialFilter
+        {
+            get
+            {
+                return uint.MaxValue;
+            }
+        }
```
Comment (alex-vazquez) [inline com.unity.render-pipelines.universal/Editor/ShaderGUI/BaseShaderGUI.cs:250]:
```suggestion
        protected virtual uint materialFilter => uint.MaxValue;
```

You can write this is shorter.

File: com.unity.render-pipelines.universal/Editor/VFXGraph/VFXURPLitMeshOutput.cs [hunk 3]
```diff
@@ -9,7 +9,15 @@ namespace UnityEditor.VFX.URP
...
     class VFXURPLitMeshOutput : VFXAbstractParticleURPLitOutput, IVFXMultiMeshOutput
     {
-        public override string name { get { return "Output Particle URP Lit Mesh"; } }
+        public override string name
+        {
+            get
+            {
+                if (shaderName != string.Empty)
+                    return $"Output Particle {shaderName} Mesh";
+                return "Output Particle URP Lit Mesh";
```
Comment (alex-vazquez) [inline com.unity.render-pipelines.universal/Editor/VFXGraph/VFXURPLitMeshOutput.cs:18]: string.IsNullOrEmpty()

File: com.unity.render-pipelines.universal/Editor/VFXGraph/VFXURPLitPlanarPrimitiveOutput.cs [hunk 4]
```diff
@@ -9,7 +9,15 @@ namespace UnityEditor.VFX.URP
...
     class VFXURPLitPlanarPrimitiveOutput : VFXAbstractParticleURPLitOutput
     {
-        public override string name { get { return "Output Particle URP Lit " + primitiveType.ToString(); } }
+        public override string name
+        {
+            get
+            {
+                if (shaderName != string.Empty)
+                    return $"Output Particle {shaderName} {primitiveType.ToString()}";
+                return "Output Particle URP Lit " + primitiveType.ToString();
```
Comment (alex-vazquez) [inline com.unity.render-pipelines.universal/Editor/VFXGraph/VFXURPLitPlanarPrimitiveOutput.cs:18]: string.IsNullOrEmpty

**Review by alex-vazquez — APPROVED** [review]

File: com.unity.render-pipelines.universal/Editor/VFXGraph/VFXURPLitPlanarPrimitiveOutput.cs (hunk 4 above)
Comment (PaulDemeulenaere) [reply com.unity.render-pipelines.universal/Editor/VFXGraph/VFXURPLitPlanarPrimitiveOutput.cs]: Fixed at 54b2a5b6d797f3880e573eda16cc30752657637e

File: com.unity.render-pipelines.universal/Editor/VFXGraph/VFXURPLitMeshOutput.cs (hunk 3 above)
Comment (PaulDemeulenaere) [reply com.unity.render-pipelines.universal/Editor/VFXGraph/VFXURPLitMeshOutput.cs]: Fixed at 54b2a5b6d797f3880e573eda16cc30752657637e

File: com.unity.render-pipelines.universal/Editor/ShaderGUI/BaseShaderGUI.cs (hunk 2 above)
Comment (PaulDemeulenaere) [reply com.unity.render-pipelines.universal/Editor/ShaderGUI/BaseShaderGUI.cs]: Fixed at 54b2a5b6d797f3880e573eda16cc30752657637e

File: com.unity.render-pipelines.high-definition/Editor/VFXGraph/Outputs/VFXLitPlanarPrimitiveOutput.cs [hunk 5]
```diff
@@ -8,7 +8,15 @@ namespace UnityEditor.VFX.HDRP
...
     class VFXLitPlanarPrimitiveOutput : VFXAbstractParticleHDRPLitOutput
     {
-        public override string name { get { return "Output Particle HDRP Lit " + primitiveType.ToString(); } }
+        public override string name
+        {
+            get
+            {
+                if (shaderName != string.Empty)
+                    return $"Output Particle {shaderName} {primitiveType.ToString()}";
+                return "Output Particle HDRP Lit " + primitiveType.ToString();
```
[the remark this answers is not in the document]
Comment (PaulDemeulenaere) [reply com.unity.render-pipelines.high-definition/Editor/VFXGraph/Outputs/VFXLitPlanarPrimitiveOutput.cs]: Fixed at 54b2a5b6d797f3880e573eda16cc30752657637e

File: com.unity.render-pipelines.high-definition/Editor/VFXGraph/Outputs/VFXLitMeshOutput.cs [hunk 6]
```diff
@@ -8,7 +8,15 @@ namespace UnityEditor.VFX.HDRP
...
     class VFXLitMeshOutput : VFXAbstractParticleHDRPLitOutput, IVFXMultiMeshOutput
     {
-        public override string name { get { return "Output Particle HDRP Lit Mesh"; } }
+        public override string name
+        {
+            get
+            {
+                if (shaderName != string.Empty)
+                    return $"Output Particle {shaderName} Mesh";
+                return "Output Particle HDRP Lit Mesh";
```
[the remark this answers is not in the document]
Comment (PaulDemeulenaere) [reply com.unity.render-pipelines.high-definition/Editor/VFXGraph/Outputs/VFXLitMeshOutput.cs]: Fixed at 54b2a5b6d797f3880e573eda16cc30752657637e

File: com.unity.render-pipelines.high-definition/Editor/Material/ShaderGraph/HDSubTarget.cs [hunk 7]
```diff
@@ -130,6 +130,12 @@ public override void Setup(ref TargetSetupContext context)
             }
         }
 
+        private static readonly string[] s_SharedTemplatePath =
+        {
+            $"{HDUtils.GetHDRenderPipelinePath()}Editor/Material/ShaderGraph/Templates/",
+            "Packages/com.unity.visualeffectgraph/Editor/ShaderGraph/Templates"
```
[the remark this answers is not in the document]
Comment (PaulDemeulenaere) [reply com.unity.render-pipelines.high-definition/Editor/Material/ShaderGraph/HDSubTarget.cs]: Fixed at 54b2a5b6d797f3880e573eda16cc30752657637e

**Review by VitaSkruibyte — APPROVED** [review]
Looking great! All the raised issues are fixed or can be looked in follow up PR. Many thanks for the PR and letting officially leave the HDRP world and switch to URP one! Can't wait to get my hands on this not as a QA, but as user. Cheers! 🟢

File: com.unity.shadergraph/Documentation~/surface-options.md [hunk 8]
```diff
@@ -49,7 +49,7 @@ This option's functionality corresponds to the command <a href="https://docs.uni
 </td></tr>
 
 
-
+<tr><td><b>Support VFX Graph</b></td><td colspan="2">This property is only available if the <a href="https://docs.unity3d.com/Packages/com.unity.visualeffectgraph@12.0/manual/">Visual Effect Graph package</a> is installed. Indicates whether this Shader Graph supports the Visual Effect Graph. If you enable this property, output contexts can use this Shader Graph to render particles. The internal setup that Shader Graph does to support visual effects happens when Unity imports the Shader Graph. This means that if you enable this property, but don't use the Shader Graph in a visual effect, there is no impact on performance. It only affects the Shader Graph import time.</td><td>Not applicable.</td></tr>
```
Comment (sarah-welton) [inline com.unity.shadergraph/Documentation~/surface-options.md:52]: Few questions: 

1. Are you saying that enabling this on a Shader Graph creates a 1:1 relationship with that Shader Graph and a VFX Graph, or just that whatever output contexts are the VFX Graph package can access the Shader Graph? 

Basically: specific VFX Graph file, or the package as a whole? 

2. This option only affects import times for importing a Shader Graph into a Visual Effect Graph?

**Review by VladNeykov — APPROVED** [review]
Looking great, no issues found, including Vita's [test doc](https://docs.google.com/document/d/1OsM3iA2O4-Zb46eUUOZRFjMaVLmWFfMkG7xLAkpkah4/edit) for reference. Thanks for implementing this, very excited for this PR to go live, it's a huge win for bringing better URP support to the VFX Graph! :)

File: com.unity.shadergraph/Documentation~/surface-options.md (hunk 8 above)
Comment (PaulDemeulenaere) [reply com.unity.shadergraph/Documentation~/surface-options.md:52]: Hello,
This documentation has been copy/past from [HDRP](https://github.com/Unity-Technologies/Graphics/blob/5e335c0df9d04dabcbe29eb22352f7c5153496f0/com.unity.render-pipelines.high-definition/Documentation~/snippets/shader-properties/support-vfx-graph.md)

> Are you saying that enabling this on a Shader Graph creates a 1:1 relationship with that Shader Graph and a VFX Graph, or just that whatever output contexts are the VFX Graph package can access the Shader Graph?

The same shaderGraph can be used in several vfx and several output, so, if I'm undersand correctly the question, it's the second option.

> This option only affects import times for importing a Shader Graph into a Visual Effect Graph?

To be honest, I'm not sure we added this precision here, typically, it indicates that you can enable it without worry about the performance impact, it only adds some metadata while importing the shadergraph to allow the potential usage in VFX.
So, actually, it affects the import time of the ShaderGraph itself in the project even if it isn't used in any VFX.

**Review by MarieGuff — APPROVED** [review]
Good to go! A few issues were found but not specific to URP integration. 

General SG/VFX issues that will need to be tackled asap:
- 4x4 Matrix compatibility (because VFX also has it)
- 3x3 and 2x2 Matrix and Sampler state properties (in SG) needs error feedback (not comaptible)
- Shader Property categories transmitted to VFX graph
**- Sorting** 
**- Reset to SG values option or automatic reset when SG values are changed**

This things were added to [Vita's Doc](https://docs.google.com/document/d/1OsM3iA2O4-Zb46eUUOZRFjMaVLmWFfMkG7xLAkpkah4/edit?usp=sharing).

**Review by cdxntchou — APPROVED** [review]
Looks good!

File: com.unity.visualeffectgraph/Editor/Compiler/VFXCodeGenerator.cs [hunk 9]
```diff
@@ -751,12 +753,23 @@ private static StringBuilder TryBuildFromShaderGraph(VFXShaderGraphParticleOutpu
             graph.ValidateGraph();
 
             // Check the validity of the shader graph (unsupported keywords or shader property usage).
-            if (!VFXLibrary.currentSRPBinder.IsGraphDataValid(graph))
+            if (VFXLibrary.currentSRPBinder == null || !VFXLibrary.currentSRPBinder.IsGraphDataValid(graph))
```
Comment (julienf-unity) [inline com.unity.visualeffectgraph/Editor/Compiler/VFXCodeGenerator.cs:756]: What's the implication of having IsGraphDataValid returning false here? Will it skip the compilation? Make it later? Try to compile anyway with bad state?

Comment (julienf-unity) [reply com.unity.visualeffectgraph/Editor/Compiler/VFXCodeGenerator.cs:756]: Also wasnt this path (compilation) not supposed to be hit at all when VFXLibrary.currentSRPBinder is null? (typically when no SRP is installed)

**Review by julienf-unity — APPROVED** [review]

File: com.unity.visualeffectgraph/Editor/Compiler/VFXCodeGenerator.cs (hunk 9 above)
Comment (PaulDemeulenaere) [reply com.unity.visualeffectgraph/Editor/Compiler/VFXCodeGenerator.cs:756]: This is a known issue we already have in HDRP, a fogbugz has been registered:
https://fogbugz.unity3d.com/f/cases/1364194/

**Review by sarah-welton — APPROVED** [review]

File: com.unity.render-pipelines.universal/Editor/ShaderGraph/Templates/ShaderPass.template [hunk 10]
```diff
@@ -114,7 +118,12 @@ $splice(GraphPixel)
 
 // --------------------------------------------------
 // Build Graph Inputs
-
+#ifdef HAVE_VFX_MODIFICATION
+#define VFX_SRP_ATTRIBUTES Attributes
```
Comment (johnpars) [inline com.unity.render-pipelines.universal/Editor/ShaderGraph/Templates/ShaderPass.template:122]: So nice to see this abstracted between pipelines!

**Review by johnpars — APPROVED** [review]
LGTM!

File: com.unity.render-pipelines.universal/Editor/ShaderGraph/Templates/ShaderPass.template (hunk 10 above)
Comment (PaulDemeulenaere) [reply com.unity.render-pipelines.universal/Editor/ShaderGraph/Templates/ShaderPass.template:122]: It was quite pleasing to end with a simple abstraction like this.
There is only one difference which remains: the `fragInputs` design is specific to HDRP and there is some spefic code in `VFXCodeGenerator.cs` to handle this indirection.

**Review by Verasl — APPROVED** [review]
LGTM

**Comment by PaulDemeulenaere** [comment]
Backport ready to be merged: https://github.com/Unity-Technologies/Graphics/pull/5687 ⏳

**Pull request merged**